ipam/paxos/paxos.go

```diff
@@ -0,0 +1,254 @@
+
+// The heart of the consensus algorithm. Return true if we have
+// changed our claims.
+func (node *Node) Think() bool {
+	our_claims := node.knows[node.id]
+
+	// The "Promise" step of Paxos: Copy the highest known
+	// promise.
+	for _, claims := range node.knows {
+		if our_claims.Promise.precedes(claims.Promise) {
+			our_claims.Promise = claims.Promise
+		}
+	}
+
+	// The "Accept Request" step of Paxos: Acting as a proposer,
+	// do we have a proposal that has been promised by a quorum?
+	//
+	// In Paxos, the "proposer" and "acceptor" roles are distinct,
+	// so in principle a node acting as a proposer could continue
+	// trying to get its proposal acccepted even after the same
+	// node as an acceptor has superseded that proposal.  But
+	// that's pointless in a gossip context: If our promise
+	// supersedes our own proposal, then anyone who hears about
+	// that promise will not accept that proposal.  So our
+	// proposal is only in the running if it is also our promise.
+	if our_claims.Promise.Proposer == node.id {
+		// Determine whether a quorum has promised, and the
+		// best previously-accepted value if there is one.
+		count := uint(0)
+		var accepted ProposalID
+		var acceptedVal AcceptedValue
+
+		for _, claims := range node.knows {
+			if claims.Promise == our_claims.Promise {
+				count++
+
+				if accepted.precedes(claims.Accepted) {
+					accepted = claims.Accepted
+					acceptedVal = claims.AcceptedVal
+				}
+			}
+		}
+
+		if count >= node.quorum {
+			if !accepted.valid() {
+				acceptedVal.Value = node.pickValue()
+				acceptedVal.Origin = our_claims.Promise
+			}
+
+			// We automatically accept our own proposal,
+			// and that's how we communicate the "accept
+			// request" to other nodes.
+			our_claims.Accepted = our_claims.Promise
+			our_claims.AcceptedVal = acceptedVal
+		}
+	}
+
+	// The "Accepted" step of Paxos: If the proposal we promised
+	// on got accepted by some other node, we accept it too.
+	for _, claims := range node.knows {
+		if claims.Accepted == our_claims.Promise {
+			our_claims.Accepted = claims.Accepted
+			our_claims.AcceptedVal = claims.AcceptedVal
+			break
+		}
+	}
+
+	if our_claims.equals(node.knows[node.id]) {
+		return false
+	}
+
+	node.knows[node.id] = our_claims
+	return true
+}
+
+// When we get to pick a value, we use the set of peer names we know
+// about.  This is not necessarily all peer names, but it is at least
+// a quorum, and so good enough for seeding the ring.
+func (node *Node) pickValue() Value {
+	val := make([]router.PeerName, len(node.knows))
+	i := 0
+	for id := range node.knows {
+		val[i] = id.Name
+		i++
+	}
+	return val
+}
+
+// Has a consensus been reached, based on the known claims of other nodes?
+func (node *Node) Consensus() (bool, AcceptedValue) {
+	counts := map[ProposalID]uint{}
+
+	for _, claims := range node.knows {
+		if claims.Accepted.valid() {
+			origin := claims.AcceptedVal.Origin
+			count := counts[origin] + 1
+			counts[origin] = count
```

Comment:
Yes, I figured it would be two lines. Not worth arguing about. So leave it as is if you like.